### PR TITLE
Refactor forgot password

### PR DESCRIPTION
**Pull request by RadoRado**

### CODE REVIEW

File: odin/authentication/apis.py [hunk 1]
```diff
@@ -57,8 +57,10 @@ def post(self, request):
 class ForgotPasswordApi(ServiceExceptionHandlerMixin, APIView):
     class Serializer(serializers.Serializer):
         user = serializers.SlugRelatedField(
-            queryset=BaseUser.objects.filter(is_active=True),
-            slug_field='email'
+            queryset=BaseUser.objects.filter(),
```
Comment (kbadova) [inline odin/authentication/apis.py]: `BaseUser.objects.all()` :thinking:

File: odin/authentication/apis.py [hunk 2]
```diff
@@ -57,8 +57,10 @@ def post(self, request):
...
+            slug_field='email',
+            error_messages={'does not exist':
```
Comment (kbadova) [inline odin/authentication/apis.py]: Better indentation:
```
error_messages: {
    'does_not_exist': 'User with that email does not exist'
}
```
:thinking:

File: odin/authentication/tests/services/test_initiate_reset_user_password.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    @patch('odin.authentication.services.send_mail')
+    def test_active_user_can_initiate_password_reset(self, mock_object):
+        self.user.is_active = True
+        self.user.save()
+
+        token = initiate_reset_user_password(user=self.user)
+
+        self.assertTrue(mock_object.called)
+        self.assertEqual(isinstance(token, PasswordResetToken), True)
+        self.assertTrue(token.voided_at is None)
```
Comment (kbadova) [inline odin/authentication/tests/services/test_initiate_reset_user_password.py]: `self.assertIsNone`

File: odin/authentication/tests/services/test_initiate_reset_user_password.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        self.assertTrue(token.used_at is None)
```
Comment (kbadova) [inline odin/authentication/tests/services/test_initiate_reset_user_password.py]: `self.assertIsNone`

File: odin/authentication/tests/services/test_reset_password.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+from odin.users.factories import BaseUserFactory
+
+from odin.authentication.services import reset_user_password
+
+from faker import Faker
+fake = Faker()
+
+
+class ResetUserpasswordTest(TestCase):
```
Comment (kbadova) [inline odin/authentication/tests/services/test_reset_password.py]: `ResetUserpasswordTests`

File: odin/authentication/tests/services/test_initiate_reset_user_password.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+from unittest.mock import patch
+
+from odin.users.factories import BaseUserFactory
+from odin.users.models import PasswordResetToken
+
+from odin.authentication.services import initiate_reset_user_password
+
+
+class InitiateResetUserPasswordTest(TestCase):
```
Comment (kbadova) [inline odin/authentication/tests/services/test_initiate_reset_user_password.py]: `InitiateResetUserPasswordTests`

File: odin/authentication/tests/services/test_reset_password.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def setUp(self):
+        self.test_password = fake.password()
+        self.user = BaseUserFactory()
+        self.user.set_password(self.test_password)
+        self.user.save()
+        self.initial_secret_key = self.user.secret_key
+        self.token = PasswordResetToken()
```
Comment (kbadova) [inline odin/authentication/tests/services/test_reset_password.py]: Can't we have `PasswordResetTokenFactory` :thinking:

File: odin/authentication/tests/services/test_reset_password.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        self.token.user = self.user
+        self.token.save()
+
+    def test_user_cannot_reset_password_with_used_token(self):
+        password = fake.password()
+        self.token.use()
```
Comment (kbadova) [inline odin/authentication/tests/services/test_reset_password.py]: We can assert the token is used:
`self.assertTrue(self.token.used)`

File: odin/authentication/tests/services/test_reset_password.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+        with self.assertRaises(ValidationError):
+            reset_user_password(token=self.token,
+                                password=password)
+
+    def test_user_cannot_reset_password_with_voided_token(self):
+        password = fake.password()
+        self.token.void()
```
Comment (kbadova) [inline odin/authentication/tests/services/test_reset_password.py]: We can assert the token is voided:
self.assertTrue(self.token.voided)

File: odin/authentication/apis.py [hunk 10]
```diff
@@ -57,11 +57,13 @@ def post(self, request):
...
+            queryset=BaseUser.objects.all(),
             slug_field='email',
-            error_messages={'does not exist':
-                            'User with that email does not exist'}
-        )
+            error_messages={
+                'does_not_exist':
+                ('User with that email does not exist')
+            }
+            )
```
Comment (kbadova) [inline odin/authentication/apis.py]: `)` One indentation <-

File: odin/authentication/tests/services/test_reset_password.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    def setUp(self):
+        self.test_password = faker.password()
+        self.token = PasswordResetTokenFactory()
+        self.user = self.token.user
+        self.user.set_password(self.test_password)
+        self.user.save()
+        self.initial_secret_key = self.user.secret_key
+
+    def test_user_cannot_reset_password_with_used_token(self):
+        password = faker.password()
```
Comment (martin056) [inline odin/authentication/tests/services/test_reset_password.py]: This can be moved to the `setUp`

File: odin/users/factories.py [hunk 12]
```diff
@@ -31,3 +31,11 @@ class ProfileFactory(factory.DjangoModelFactory):
 
     class Meta:
         model = Profile
+
+
+class PasswordResetTokenFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = PasswordResetToken
+
+    token = uuid.uuid4()
```
Comment (martin056) [inline odin/users/factories.py]: Use `faker.uuid4()`

Comment (yaggul) [reply odin/users/factories.py]: we cannot. The model requires UUIDField that is UUID object in the DB. faker.uuid4() returns uuid4 as str
i have already tested it.

**Pull request merged**